app/service/data_svc.py

```diff
@@ -19,6 +19,25 @@
 from app.utility.base_service import BaseService
 
 Adjustment = namedtuple('Adjustment', 'ability_id trait value offset')
+history = defaultdict(set)
+
+
+def track_recursion(f):
+    """
+    decorator to track recursive calls in stateless functions to prevent infinite loops
+    """
+    function_history = history[f.__name__]
+
+    async def call(*args, **kwargs):
+        key = str({'args': args, 'kwargs': kwargs})
+        if key in function_history:
+            function_history.clear()
+            raise Exception(f'infinite loop detected, function:{f.__name__}, inputs:{key}')
```

Comment:
anything catch this exception or is it game over?

Reply:
yup, currently it is caught in the `_load_adversaries(...)` call as it is running through adversaries so that a failure on one wont prevent all prior ones from loading but will still allow us to break from the loop